libbeat/processors/actions/decode_json_fields.go

```diff
@@ -0,0 +1,101 @@
+package actions
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/jsontransform"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+	"github.com/pkg/errors"
+)
+
+type decodeJSONFields struct {
+	Fields []string
+}
+
+var debug = logp.MakeDebug("filters")
+
+func init() {
+	processors.RegisterPlugin("decode_json_fields", configChecked(newDecodeJSONFields,
+		requireFields("fields"), allowedFields("fields", "when")))
+}
+
+func newDecodeJSONFields(c common.Config) (processors.Processor, error) {
+	config := struct {
+		Fields []string `config:"fields"`
+	}{}
+	err := c.Unpack(&config)
+	if err != nil {
+		logp.Warn("Error unpacking config for decode_json_fields")
+		return nil, fmt.Errorf("fail to unpack the decode_json_fields configuration: %s", err)
+	}
+
+	f := decodeJSONFields{Fields: config.Fields}
+	return f, nil
+}
+
+func (f decodeJSONFields) Run(event common.MapStr) (common.MapStr, error) {
+	var errs []string
+
+	for _, field := range f.Fields {
+		data, err := event.GetValue(field)
+		if err != nil && errors.Cause(err) != common.ErrKeyNotFound {
+			debug("Error trying to GetValue for field : %s in event : %v", field, event)
+			errs = append(errs, err.Error())
+			continue
+		}
+		text, ok := data.(string)
+		if ok {
+			var output map[string]interface{}
+			err := unmarshal([]byte(text), &output)
+			if err != nil {
+				debug("Error trying to unmarshal %s", event[field])
+				errs = append(errs, err.Error())
+				continue
+			}
+
+			_, err = event.Put(field, output)
+			if err != nil {
+				debug("Error trying to Put value %v for field : %s", output, field)
+				errs = append(errs, err.Error())
+				continue
+			}
+		}
+	}
+
+	return event, fmt.Errorf(strings.Join(errs, ", "))
+}
+
+// unmarshal is equivalent with json.Unmarshal but it converts numbers
+// to int64 where possible, instead of using always float64.
+func unmarshal(text []byte, fields *map[string]interface{}) error {
+	dec := json.NewDecoder(bytes.NewReader(text))
+	dec.UseNumber()
+	err := dec.Decode(fields)
+	if err != nil {
+		return err
+	}
+
+	//Iterate through all the fields to perform deep parsing
+	for k, v := range *fields {
+		switch vv := v.(type) {
+		case string:
+			var output map[string]interface{}
+			sErr := unmarshal([]byte(vv), &output)
+			if sErr == nil {
+				(*fields)[k] = output
+			}
+		}
+	}
+
+	jsontransform.TransformNumbers(*fields)
+	return nil
+}
```

Comment:
thinking a little more about `TransformNumbers` and `unmarshal`, I wonder why wen need to put `TransformNumbers` at the end. By moving `TransformNumber` before the deep-parsing stuff, the initial decoding would be exactly similar to the filebeat reader.

```
func DecodeJson(text []byte, to interface{}) {
  dec := json.NewDecoder(bytes.NewReader(text))
  dec.UseNumber()
  err := dec.Decode(to)
  if err != nil {

  if m, ok := to.(*map[string]interface{}); ok {
    jsontransform.TransformNumbers(common.MapStr(m))
  }
  return err
}
```

Then (extending unmarshal a little), unmarshal can be generalized a little:

```
func unmarshal(maxDepth int, text []byte, fields interface{}, processArray bool) error {
  if err := DecodeJson(text, fields); err != nil {
    return err
  }

  maxDepth--
  if maxDepth == 0 {
    return nil
  }

  tryUnmarshal := func(v interface{}) (interface{}, bool) {
    str, isString := v.(string)
    if !isString {
      return v, false
    }

    var tmp interface{}
    err := unmarshal(maxDepth, []byte(str), &tmp, processArray)
    if err != nil {
      return v, false
    }

    return tmp, true
  }

  // try to deep unmarshal fields
  switch O := fields.(type); {
  case *map[string]interface{}:
    for k, v := range O {
      if decoded, ok := tryUnmarshal(v); ok {
        O[k] = decoded
      }
    }

  // We want to process arrays here
  case []interface{}:
    if !processArray {
      break
    }

    for i, v := range O {
      if decoded, ok := tryUnmarshal(v); ok {
        O[i] = decoded
      }
    }
  }
}
```

As you can see, I added a maximum parsing depth. Not sure we really want to parse fully recursively here. I'd use `maxDepth=1` by default (Make it configurable). More parsing can be done by configuring a second processor continue parsing on already parsed field.
I also added optional array support, plus lifted the requirement for JSON Objects parsing only.

Having DecodeJson, there is no need for exporting jsontransform as is (but I'd keep jsontransform package for now, as other modules in beats might make use of it).

What do you think?